design/aws-ecs-scheduled-ecs-task-construct.md

```diff
@@ -0,0 +1,141 @@
+   */
+  readonly desiredTaskCount?: number;
+
+  /**
+   * The hard limit (in MiB) of memory to present to the container.
+   *
+   * If your container attempts to exceed the allocated memory, the container
+   * is terminated.
+   *
+   * @default 256
+   */
+  readonly memoryLimitMiB?: number;
+
+  /**
+   * The schedule or rate (frequency) that determines when CloudWatch Events
+   * runs the rule. For more information, see Schedule Expression Syntax for
+   * Rules in the Amazon CloudWatch User Guide.
+   *
+   * @see http://docs.aws.amazon.com/AmazonCloudWatch/latest/events/ScheduledEvents.html
+   *
+   * You must specify this property, the `eventPattern` property, or both.
+   */
+  readonly scheduleExpression?: string;
+
+  /**
+   * Whether to create an AWS log driver to enable logging.
+   *
+   * @default true
+   */
+  readonly enableLogging?: boolean;
+
+  // Container Overrides
+
+  /**
+   * The CMD value to pass to the container.
+   *
+   * If you provide a shell command as a single string, you have to quote command-line arguments.
+   *
+   * @default CMD value built into container image
+   */
+  readonly command?: string[];
+
+  /**
+   * The minimum number of CPU units to reserve for the container passed as a container override.
+   */
+  readonly cpu?: number;
+
+  /**
+   * Environment variables to pass to the container as a container override.
+   *
+   * @default No environment variables
+   */
+  readonly environment?: { [key: string]: string };
+
+  /**
+   * Resource requirements to pass to the container as a container override. 
+   * The type and amount of a resource to assign to a container. The only supported resource is a GPU. 
+   * For more information, see http://docs.aws.amazon.com/AmazonECS/latest/developerguide/ecs-gpu.html 
+   *
+   * @default No resource requirements
+   */
+  readonly resourceRequirements?: { [key: string]: string };
+}
+```
+
+The `ScheduledEc2Task` construct will use the following existing constructs:
+
+* Ec2TaskDefinition - To create a Task Definition for the container to start
+* Ec2EventRuleTarget - The target of the aws event
+* EventRule - To describe the event trigger (in this case, a scheduled run)
+
+An example use case to create a task that is scheduled to run every minute and environment variables passed in as overrides to the container:
+```ts
+// Create the vpc and cluster used by the scheduled task
+const vpc = new ec2.VpcNetwork(stack, 'Vpc', { maxAZs: 1 });
+const cluster = new ecs.Cluster(stack, 'EcsCluster', { vpc });
+cluster.addCapacity('DefaultAutoScalingGroup', {
+  instanceType: new ec2.InstanceType('t2.micro')
+});
+
+// Create the scheduled task with environment variables passed in as overrides
+new ScheduledEc2Task(stack, 'ScheduledEc2Task', {
+  cluster,
+  image: ecs.ContainerImage.fromRegistry('amazon/amazon-ecs-sample'),
+  memoryLimitMiB: 512,
+  scheduleExpression: 'rate(1 minute)',
+  environment: { name: 'TRIGGER', value: 'CloudWatch Events' }
```

Comment:
It's not clear from this example whether these env settings will be registered with the task definition or applied as overrides when the schedule task runs. What do you think about (optionally?) taking in a customer-defined [Ec2]TaskDefinition construct to run? That would allow customers to run multiple containers and already-existing* task definitions with this L3.

*Is it possible to "import" an existing task definition into an L2?

Reply:
Passing a TaskDefinition object seems more valuable in the CloudWatch Events Rule Target L2, so that the task def ARN parameter gets populated automatically.  In fact, I think that should already be supported?  So that the ecsParameters value ends up being something like "!Ref MyTaskDefinition123"

My goal for L3s is that they are oriented towards the user only creating a single object that encapsulates everything.  Everything else should be possible at the L2 level.

Reply:
Ah, if that's the intention behind L3s (that one obj does all the work), and scheduled task support for  existing Task Defs is something customers can already do with L2s, then I understand why we're taking in an image directly here. Customers can bail out and use L2s if they need more specifics.